feat: add tracing to worker and proxy

SantiagoPittella opened the pull request:
this PR is part of #1004

bobbinth commented on bin/tx-prover/src/commands/proxy.rs:
```diff
@@ -18,6 +18,7 @@ impl StartProxy {
     ///
     /// This method will first read the config file to get the list of workers to start. It will
     /// then start a proxy with each worker as a backend.
+    #[tracing::instrument(target = "miden-tx-prover", name = "proxy:execute")]
```
We should now be able to use constants here (e.g., `MIDEN_TX_PROVER`) instead of string literals.

bobbinth commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -333,6 +343,7 @@ impl RequestContext {
 /// This wrapper is used to implement the ProxyHttp trait for Arc<LoadBalancer>.
 /// This is necessary because we want to share the load balancer between the proxy server and the
 /// health check background service.
+#[derive(Debug)]
 pub struct LoadBalancer(pub Arc<LoadBalancerState>);
```
Not from this PR, but we should update section header on line 338 above.

SantiagoPittella replied:
I'm adding this task to the error handling issue.

bobbinth commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -535,6 +551,135 @@ impl ProxyHttp for LoadBalancer {
         }
     }
+
+    // The following methods are a copy of the default implementation defined in the trait, but
+    // with tracing instrumentation.
```
Could we add a brief explanation of why we need these methods implemented?

bobbinth commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -300,22 +305,27 @@ static QUEUE: Lazy<RequestQueue> = Lazy::new(RequestQueue::new);
-            request_id: Uuid::new_v4(),
+            request_id,
             worker: None,
+            parent_span: info_span!("proxy:new_request", request_id = request_id.to_string()),
```
Question: why don't we need to specify `target` both here and in other places in this file?

SantiagoPittella replied:
We **should** be using `target` here. For the "sons" (the traces that we create in each method of `HttpProxy`) it is not needed since there it gets inherited.

bobbinth commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -535,6 +551,135 @@ impl ProxyHttp for LoadBalancer {
+    where
+        Self::CTX: Send + Sync,
+    {
+        let server_session = session.as_mut();
+        let code = match e.etype() {
+            HTTPStatus(code) => *code,
+            _ => {
+                match e.esource() {
+                    ErrorSource::Upstream => 502,
+                    ErrorSource::Downstream => {
+                        match e.etype() {
+                            WriteError | ReadError | ConnectionClosed => {
+                                /* conn already dead */
+                                0
+                            },
+                            _ => 400,
+                        }
+                    },
+                    ErrorSource::Internal | ErrorSource::Unset => 500,
+                }
+            },
+        };
+        if code > 0 {
+            server_session.respond_error(code).await
+        }
+        code
```
I'm assuming that this is just a copy of a default implementation, right?

SantiagoPittella replied:
Yes, it is. I'm already replacing it with `<Trait>::<method>` thought.

bobbinth commented on bin/tx-prover/src/utils.rs:
```diff
@@ -1,20 +1,71 @@
-        std::env::set_var("RUST_LOG", "info"); // Default to 'info' level
-    }
-    tracing_subscriber::fmt::init();
+// Construct TracerProvider for OpenTelemetryLayer
+pub(crate) fn init_tracer_provider() -> TracerProvider {
```
nit: let's use `///` for doc comments (here and in other places in this file).

Also, could we add more details about how we configure the tracing provider? For example, why do we need to add ID generator, what does `with_sampler()` do etc.

bobbinth commented on bin/tx-prover/src/utils.rs:
```diff
@@ -1,20 +1,71 @@
+    )
 }
 
+// Setup tracing subscriber
+pub(crate) fn setup_tracing(provider: TracerProvider) -> Result<(), String> {
```
Similar to the previous comment - could we add more details about what this function does?

Also, what's the motivation for have this and `init_tracer_provider()` as two separate functions? It seems like one is called right after the other. Should we combine them into one function?

bobbinth commented on bin/tx-prover/src/utils.rs:
```diff
@@ -1,20 +1,71 @@
+use tracing::Level;
+use tracing_subscriber::{layer::SubscriberExt, Registry};
+
+pub const TRACING_TARGET_NAME: &str = "miden-tx-prover";
```
nit: I maybe would still call it `MIDEN_TX_PROVER`.

bobbinth commented on bin/tx-prover/Cargo.toml:
```diff
@@ -32,12 +32,19 @@ tonic = { version = "0.12", default-features = false, features = ["prost", "code
+opentelemetry = { version = "0.27", features = ["metrics", "trace"] }
+opentelemetry-otlp = { version = "0.27", features = ["grpc-tonic"] }
+opentelemetry_sdk = { version = "0.27", features = ["metrics", "rt-tokio"] }
+opentelemetry-semantic-conventions = "0.27.0"
+opentelemetry-jaeger = "0.22.0"
```
nit: I would get rid of the patch versions.

bobbinth commented on bin/tx-prover/README.md:
```diff
@@ -114,6 +114,16 @@ The proxy service uses this health check to determine if a worker is available t
 
 Both the worker and the proxy will use the `info` log level by default, but it can be changed by setting the `RUST_LOG` environment variable.
 
+## Traces
```
I would maybe combine this and logging into one section (or make this a sub-section of logging?).

Also, I would add more details here. For example, it is not clear where tracing/logging info is written to. Is it `stdout`? Is it some logging file? Somewhere else? It is also not clear whether there is a way to not use Jaeger (or maybe use something else) to view the logs.

Basically, a bit more context about how tracing/logging works would be helpful.

bobbinth commented on bin/tx-prover/README.md:
````diff
@@ -114,6 +114,16 @@ The proxy service uses this health check to determine if a worker is available t
+The service uses the `tracing` crate for structured logging and tracing. Traces are enabled by default, and uses opentelemetry to export traces to a Jaeger instance. The traces can be visualized using the Jaeger UI, which can be used by running:
+
+```bash
+docker run -d -p4317:4317 -p16686:16686 jaegertracing/all-in-one:latest
````
This assumes that we have `docker` installed on the machine, right? If so, I would mention this.

Also, are there alternative ways to do this? We don't need to describe them but if there is a link to how to do it w/o Docker, I'd include it.

bobbinth reviewed (commented):
Looks good! Thank you! I left some comments inline - most doc-related, but would be good for @igamigo and @Mirko-von-Leipzig to take a look as well.

igamigo commented on bin/tx-prover/src/utils.rs:
```diff
@@ -1,20 +1,71 @@
+    let exporter = create_span_exporter();
+
+    TracerProvider::builder()
+        .with_sampler(Sampler::ParentBased(Box::new(Sampler::TraceIdRatioBased(1.0))))
```
Not necessarily for this PR: Not sure if this will impact performance on production since it will depends on traces, but maybe this `1.0` could be an env var or configurable somehow.

SantiagoPittella replied:
That would be great. It might be good to determine if the tracing/logging/metrics related configuration should be done through env vars like `RUST_LOG=<level>` or use a configuration file.

Mirko-von-Leipzig replied:
Why not just make it a cli arg with env var support. clap supports this out the box, and personally I really dislike having these config files 😅 They just make deployment a nightmare.

SantiagoPittella replied:
Opened up [this issue](https://github.com/0xPolygonMiden/miden-base/issues/1026) to tackle this.

igamigo commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -535,6 +551,135 @@ impl ProxyHttp for LoadBalancer {
         }
     }
+
+    // The following methods are a copy of the default implementation defined in the trait, but
+    // with tracing instrumentation.
+    // ============================================================================================
```
I think for default implementations, we should just be able to use `ProxyHttp::{method_name}` instead of copying them (unless there's something I'm missing).

SantiagoPittella replied:
Oh, I did not know that using ProxyHttp::{method} calls the default implementation. That is so much better. Thanks!

igamigo replied:
Looks like I was having hallucinations and this is in fact not possible

igamigo commented on bin/tx-prover/src/api/mod.rs:
```diff
@@ -30,10 +32,18 @@ pub struct ProverRpcApi {
         request: Request<ProveTransactionRequest>,
     ) -> Result<Response<ProveTransactionResponse>, tonic::Status> {
+        debug!(request = ?request, "Processing reply");
         info!("Received request to prove transaction");
```
Since this function has the instrument macro, does this provide anything valuable? Perhaps if looking at the logs on stdout?

SantiagoPittella replied:
It does not make much sense to keep it. If looking at stdout out we already the log from inside the prover:
```
prover:prove_transaction:prove_program: miden_prover: Generated execution trace ...
```

I'm removing it and also changing the instrumentation. In particular, I'm changing the log level of the return vale log from `info` to `debug`. 

Currently, using `info` logs:
```
2024-12-16T17:45:13.526435Z  INFO prover:prove_transaction: miden-tx-prover: return=Response { metadata: MetadataMap { headers: {} }, message: ProveTransactionResponse { proven_transaction: [187, 199, 33, ... ] }
```
And `proven_transaction` is quite big to have it on by default.

SantiagoPittella commented:
@bobbinth @igamigo I answer a couple of your questions on the code/documentation itself. In particular the ones about tracing setup and Jaeger. Let me know if something was left unattended.

bobbinth commented on bin/tx-prover/README.md:
```diff
@@ -110,9 +110,39 @@ The worker service implements the [gRPC Health Check](https://grpc.io/docs/guide
+
+### Configuring Tracing
+
+- **Trace Exporter**: Traces are exported to a Jaeger instance using the OpenTelemetry protocol. The Jaeger instance can be run locally using Docker (see below).
```
I'm still not quite clear on how this works. I'm assuming the service sends traces some local port - right? Does it have to be a Jaeger instance (i.e., there is no mentioning of Jaeger when we configure OpenTelemetry tracing)? How does it know which instance to use? If we have 2 Jaeger instances, can we specify somehow which one to use? Does it always have to be Jaeger or can we use some other OpenTelemetry subscriber?

And what if the instance is not there? Would the service somehow detect this and discard all the traces?

bobbinth commented on bin/tx-prover/src/utils.rs:
```diff
@@ -1,18 +1,109 @@
+/// ### Returns
+/// - `Ok(TracerProvider)`: If the tracer provider is successfully initialized.
+/// - `Err(String)`: If the OTLP exporter fails to initialize, an error message is returned.
+pub(crate) fn init_tracer_provider() -> Result<TracerProvider, String> {
```
Does this function need to be `pub(crate)`? It seems like we use it only internally in this module.

If the above is correct, it is still not clear to me why we don't just combine it with `setup_tracing()` function (i.e., move the code here into `setup_tracing()` function).

bobbinth commented on bin/tx-prover/README.md:
````diff
@@ -110,9 +110,39 @@ The worker service implements the [gRPC Health Check](https://grpc.io/docs/guide
+## Logging and Tracing
 
-Both the worker and the proxy will use the `info` log level by default, but it can be changed by setting the `RUST_LOG` environment variable.
+The service uses the [`tracing`](https://docs.rs/tracing/latest/tracing/) crate for both logging and distributed tracing, providing structured, high-performance logs and trace data. By default, logs are written to `stdout`, and traces are exported to a Jaeger instance via OpenTelemetry.
+
+### Configuring Logging
+
+- **Default Log Level**: Both the worker and proxy use the `info` log level by default.
+- **Custom Log Levels**: You can configure log levels by setting the `RUST_LOG` environment variable. For example:
+    ```
+    export RUST_LOG=debug
+    ```
+- **Log Destination**: Logs are output to `stdout` by default.
````
I'm not sure we need this much boilerplate here. I'd write these as simple sentences and paragraphs. For example:

-----

The service uses the [`tracing`](https://docs.rs/tracing/latest/tracing/) crate for both logging and distributed tracing, providing structured, high-performance logs and trace data.

By default, logs are written to `stdout` and the default logging level is `info`. This can be changed via the `RUST_LOG` environment variable. For example:
```
export RUST_LOG=debug
```

For tracing, we use OpenTelemetry protocol. By default, traces are exported to ...

bobbinth commented on bin/tx-prover/README.md:
```diff
@@ -110,9 +110,39 @@ The worker service implements the [gRPC Health Check](https://grpc.io/docs/guide
+### Configuring Tracing
+
+- **Trace Exporter**: Traces are exported to a Jaeger instance using the OpenTelemetry protocol. The Jaeger instance can be run locally using Docker (see below).
+- **Trace Visualization**: Once traces are exported, they can be visualized in the Jaeger UI. You can access the Jaeger UI at `http://localhost:16686/` if running locally.
```
Is `16686` here something that Jaeger defines? Or is it the default port for OpenTelemetry?

bobbinth reviewed (commented):
Thank you! Looks good! I left a few more comments inline.

Mirko-von-Leipzig commented on bin/tx-prover/src/api/mod.rs:
```diff
@@ -30,12 +32,17 @@ pub struct ProverRpcApi {
+    #[instrument(
+        target = MIDEN_TX_PROVER,
+        name = "prover:prove_transaction",
+        skip_all,
```
I would include at least the transaction ID

Mirko-von-Leipzig commented on bin/tx-prover/src/commands/proxy.rs:
```diff
@@ -18,6 +21,7 @@ impl StartProxy {
     ///
     /// This method will first read the config file to get the list of workers to start. It will
     /// then start a proxy with each worker as a backend.
+    #[tracing::instrument(target = MIDEN_TX_PROVER, name = "proxy:execute")]
```
Should this get a different proxy `target`?

SantiagoPittella replied:
Not sure, my idea was to group both services under the same target in order to be able to eventually join traces (adding the worker trace to the proxy essentially). Is that ok? Maybe there is other way.

Mirko-von-Leipzig replied:
I'm unsure of how e.g. jaeger wants it, but I imagine one needs a common request UUID of sorts in order to aggregate a single request as it flows through.

Would the transaction ID not be good enough if we add it as a top level span in each process?

A potential issue would be someone sending the same transaction multiple times; in which case we would need to attach a more unique request identifier which is usually communicated using the `X-Request-ID` http extension header. Though this feels unnecessary at this stage, and probably we only care about this on a transaction ID level.

You already have more practical experience with jaeger though, so maybe what I'm thinking isn't applicable.

SantiagoPittella replied:
Both from the proxy and worker perspective I think that eventually adding a `X-Request-ID` header is the best option, using the UUID created in the proxy.

Mirko-von-Leipzig replied:
Could you create a follow-up issue for this? Likely we'll also want to do this in the node, and maybe even in the client.

SantiagoPittella replied:
Done #1031

SantiagoPittella commented:
@bobbinth I simplified the readme following your suggestion above and added a small part about how to change the exporter endpoint in case we want to use another solution.

Also, I ended up changing the calls `HttpProxy::{method}` because it wasn't working as expected. Contrary to my belief, it was calling the LoadBalancer implementation instead of the trait one. The workaround that I implemented is creating a dummy structure to implemented the trait too.

bobbinth commented on bin/tx-prover/README.md:
````diff
@@ -110,9 +110,30 @@ The worker service implements the [gRPC Health Check](https://grpc.io/docs/guide
+export RUST_LOG=debug
+```
+
+For tracing, we use OpenTelemetry protocol. By default, traces are exported to a Jaeger instance. To visualize them locally using [Docker](https://www.docker.com/), run:
````
Based on my current understanding, Jaeger is not really the default, right? Maybe saying something like this would be more accurate?:

------

For tracing, we use OpenTelemetry protocol. By default, traces are exported to the endpoint specified by `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable. To consume and visualize these traces we can use Jaeger or any other OpenTelemetry compatible consumer.

The simplest way to install Jaeger is by using a [Docker](https://www.docker.com/) container ...

bobbinth reviewed (approved):
Looks good! Thank you! I left one last doc-related comment inline.

SantiagoPittella commented:
I'm currently working on rebasing this branch with `next`. I don't know if @Mirko-von-Leipzig wanted to perform a review of it before merging it. I plan to merge it at end of day today.

Mirko-von-Leipzig commented:
> I'm currently working on rebasing this branch with `next`. I don't know if @Mirko-von-Leipzig wanted to perform a review of it before merging it. I plan to merge it at end of day today.

I'll do a more thorough review now.

Mirko-von-Leipzig commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 
 ### Changes
 
+- Added tracing to the `miden-tx-prover` CLI (#1014).
```
I think we add these to the bottom of the list i.e. chronologically

Mirko-von-Leipzig commented on bin/tx-prover/Cargo.toml:
```diff
@@ -32,12 +32,19 @@ tonic = { version = "0.12", default-features = false, features = ["prost", "code
 pingora-core = "0.4"
 pingora-proxy = "0.4"
 pingora-limits = "0.4"
+opentelemetry = { version = "0.27", features = ["metrics", "trace"] }
+opentelemetry-otlp = { version = "0.27", features = ["grpc-tonic"] }
+opentelemetry_sdk = { version = "0.27", features = ["metrics", "rt-tokio"] }
+opentelemetry-semantic-conventions = "0.27"
+opentelemetry-jaeger = "0.22"
+tracing-opentelemetry = "0.28"
```
Side comment: You'd think they'd re-export them at least..

Mirko-von-Leipzig commented on bin/tx-prover/README.md:
```diff
@@ -110,9 +110,26 @@ The worker service implements the [gRPC Health Check](https://grpc.io/docs/guide
-Both the worker and the proxy will use the `info` log level by default, but it can be changed by setting the `RUST_LOG` environment variable.
+The service uses the [`tracing`](https://docs.rs/tracing/latest/tracing/) crate for both logging and distributed tracing, providing structured, high-performance logs and trace data.
+
+By default, logs are written to `stdout` and the default logging level is `info`. This can be changed via the `RUST_LOG` environment variable. For example:
```
```suggestion
By default, logs are written to `stdout` and the default logging level is `info`. This can be changed via the `RUST_LOG` environment variable. For example:

```

Mirko-von-Leipzig commented on bin/tx-prover/src/proxy/mod.rs:
```diff
@@ -535,6 +551,146 @@ impl ProxyHttp for LoadBalancer {
+        ProxyHttpDefaultImpl.early_request_filter(_session, &mut ()).await
+    }
+
+    #[tracing::instrument(name = "proxy:connected_to_upstream", parent = &ctx.parent_span, skip(_session, _sock, _fd))]
```
Should we not be skipping `all` in these? Or do we want to log context etc?

SantiagoPittella replied:
We should keep context here, but I'm removing all other params

Mirko-von-Leipzig reviewed (approved):
I think this is a very good first stab. Once we have a better feeling for how dashboards, traces and logs look we can refine what we in/exclude.
